src/Types/Infer/App.hs

```diff
@@ -0,0 +1,224 @@
+{-# LANGUAGE FlexibleContexts, TupleSections, ScopedTypeVariables,
+   TypeFamilies, CPP, StandaloneDeriving, UndecidableInstances #-}
+module Types.Infer.App (inferApps) where
+
+import Prelude
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import Data.Traversable
+import Data.Foldable
+import Data.Spanned
+import Data.Reason
+import Data.Triple
+import Data.Maybe
+
+import Control.Monad.Infer
+import Control.Lens
+
+import Syntax.Subst
+import Syntax.Var
+import Syntax
+
+import Types.Infer.Builtin
+import {-# SOURCE #-} Types.Infer
+import Types.Kinds
+import Types.Unify
+
+import Text.Pretty.Semantic
+
+
+#ifdef TRACE_TC
+import Debug.Trace
+#endif
+
+-- Infer an application, supporting "Quick Look" impredicativity: We
+-- consider all of the arguments to a function application (or
+-- operator) together, then do two passes over them:
+--
+--  * The first pass is what we call a 'quick look'. Here, we try to
+--  infer polymorphic type instantiations for some of the function's
+--  arguments. This pass only considers "simple" arguments: Variables
+--  (for now).
+--
+--  * The second pass is the real use of 'checks' that we've always done
+--  for arguments.
+inferApps :: forall m. MonadInfer Typed m => Expr Desugared -> Maybe (Type Typed) -> m (Expr Typed, Type Typed)
+inferApps exp expected =
+  do
+    let ((ExprArg function, _):arguments) = reverse $ spine exp
+    (function, function_t) <- infer' function
+    function_t <- refresh function_t
+
+#ifdef TRACE_TC
+    traceM ("function type: " ++ displayS (pretty function_t))
+#endif
+
+    ((ql_sub, quantifiers), cs) <-
+      censor (const mempty) . listen $
+        quickLook function_t =<< for arguments (\a@(x, y) -> (x, y,) <$> inferQL a)
+
+    quantifiers <- pure (applyQ ql_sub quantifiers)
+
+    let fake_ty_con = TyCon (TgInternal T.empty)
+    for_ (Map.toList ql_sub) $ \(var, tau) ->
+      -- Here we need these to be guarded, so they need to be in the RHS
+      -- of a type application. Since we don't have a suitable type, we
+      -- use an internal TyCon with an invisible name. (Evil!)
+      unify (becauseExp exp) (TyApp fake_ty_con (TyVar var)) (TyApp fake_ty_con tau)
+
+    tell cs
+
+    r_ql_sub <- case expected of
+      Just tau@(TyApps (TyCon _) (_:_)) | result@(TyApps (TyCon _) (_:_)) <- getQuantR quantifiers -> do
+-- Pushing down the result type into the quick look substitution
+-- is only sound if both are /guarded/. Do note that missing a
+-- substitution here isn't unsound, it'll just lead to wonky inference.
+        _ <- subsumes (becauseExp exp) result (apply ql_sub tau)
+        pure $ fromMaybe mempty (unifyPure result tau)
+      _ -> pure mempty
+
+#ifdef TRACE_TC
```

Comment:
Is it worth refactoring the trace code from `Types.Unify` into a `Types.Trace` module? Then we can share it everywhere, rather than using the CPP in some places, and not in others.

Reply:
No: this lets us individually control which modules to trace. Kinda sad, though